### ISSUE TITLE
Rule Change: Make max-lines-per-function only reports lines that exceed the limit, not all

### ISSUE THREAD

**Issue by alvis** [eslint/eslint#15098]
### What rule do you want to change?

max-lines-per-function

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/* eslint max-lines-per-function: ["error", {"max": 3}] */
export function bulkyFunction() {
  // line 1
  // line 2
  // line 3
}
```

### What does the rule currently do for this code?

Currently it will report the range of line numbers from the beginning of the function to the very last line of it.

### What will the rule do after it's changed?

Following the same spirit of `max-lines` implemented in #13318, we should only report those lines which exceed the limit.
With this example, it should only report line 4 and 5.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

The current default is annoying because once the function exceeds even a single line, the whole function is highlighted by the editor and other important errors may be suppressed as a result.

**Comment by nzakas** [comment]
I’d say this is a bug given the different behavior in max-lines, so I’m in favor of bringing this rule into alignment.

**Comment by btmills** [comment]
Also in favor. #13318 updated `max-lines` as part of the broader highlight location updates in in #12334, and those were done as semver-minor enhancements, so I'd say this should be done that way as well instead of as a semver-patch bug fix.

**Comment by nzakas** [comment]
Fair enough. @alvis would you be interested in submitting a pull request for this?

**Comment by LilCthulu** [comment]
I would love to take this on as first contribution if possible!

**Comment by alvis** [comment]
@LilCthulu I'd appreciate it very much if you can help ❤️

**Comment by nzakas** [comment]
@LilCthulu happy to have your help. Please post back here if you have any questions or need some pointers.

**Comment by LilCthulu** [comment]
@nzakas @alvis Awesome! I'll get started right away. Any specific file to start at? Also anything general to know before I dive in?

**Comment by nzakas** [comment]
This is the main file to look at:
https://github.com/eslint/eslint/blob/master/lib/rules/max-lines-per-function.js

You’ll also need to update the tests:
https://github.com/eslint/eslint/blob/master/tests/lib/rules/max-lines-per-function.js

The pull request making a similar change to max-lines might be useful as an example:
https://github.com/eslint/eslint/pull/13318/files

**Comment by The-x-Theorist** [comment]
I'd like to make my first open source contribution by updating this rule. 😄

**Comment by LilCthulu** [comment]
@nzakas  I am indeed sorry for being slow! Had a bunch of stuff going on.

**Comment by nzakas** [comment]
@LilCthulu thats okay, we just need regular updates because there are others interested in implementing this. If you can’t get to this soon, please let us know so we can let someone else work on it. 

@Gautam-Arora24 right now @LilCthulu is working on this and @The-x-Theorist is next line.

**Comment by LilCthulu** [comment]
@nzakas Sorry for the delay it is fine if someone else grabs this

**Comment by nzakas** [comment]
@Gautam-Arora24 @The-x-Theorist has already submitted a pull request.

**Comment by nzakas** [comment]
Here it is: https://github.com/eslint/eslint/pull/15140

**Comment by mdjermanovic** [comment]
> Since the input is `node` in `max-lines-per-function`, we should be able use `// eslint-disable-line max-lines-per-functions` just right before the function as in the past.

Location of `// eslint-disable-line` relates to the reported `line` in the [`LintMessage`](https://eslint.org/docs/developer-guide/nodejs-api#-lintmessage-type). If `loc` is present in `context.report({})`, `line` will be `loc.start.line`. Only if `loc` is not present, `line` will be `node.loc.start.line`. So, when we add `loc`, `node` will not matter for disable comments.

For example, this disable comment:

```js
/* eslint max-lines-per-function: ["error", 5] */

function one() { // eslint-disable-line max-lines-per-function
    two;
    three;
    four;
    five;
    six;
}
```

would have to be relocated here:

```js
/* eslint max-lines-per-function: ["error", 5] */

function one() {
    two;
    three;
    four;
    five;
    six; // eslint-disable-line max-lines-per-function
}

```

Then, if anything new is added to the function before that line, the location of the comment must be changed:

```js
/* eslint max-lines-per-function: ["error", 5] */

function one() {
    two;
    three;
    somethingNew;
    four;
    five; // eslint-disable-line max-lines-per-function
    six; 
}
```

**Comment by nzakas** [comment]
Can’t people just use a disable comment before the function and then enable after? I don’t think we need to guarantee that every disable directive works in every situation.

**Comment by mdjermanovic** [comment]
> Can’t people just use a disable comment before the function and then enable after?

Yes, that will work. A downside is that it also disables this rule in inner functions. Anyway, since everyone is in favor of this proposal, I agree that we try it out.

Question: do we want to change the error message (as in https://github.com/eslint/eslint/pull/15140), or shall we keep the existing one?

**Comment by nzakas** [comment]
I’m fine either way. There’s something nice about doing the math for users, though.

**Comment by btmills** [comment]
Thinking about this from the user's perspective, seeing editor squigglies only on the excess lines would be much nicer than squigglying the whole function. The tradeoff that disable-(next-)line comments will have to be moved might be seen as a feature: previously, disabling this rule for a function was binary. Now, someone could disable it on just the n+1th line, but adding an n+2nd line would still error. Block comments are still available to disable the rule for an entire function.

**Comment by ljharb** [comment]
https://github.com/eslint/eslint/pull/15140#issuecomment-985969953

> This seems like a bit of a breaking change.
>
> It's causing my eslint-disable comments - that have always been positioned at the top of the function - to be considered unused, and instead the random line that happens to break the threshold is reported on.

**Comment by mdjermanovic** [comment]
> It's causing my eslint-disable comments - that have always been positioned at the top of the function - to be considered unused, and instead the random line that happens to break the threshold is reported on.

Can you move the `eslint-disable-line` (or `eslint-disable-next-line`) comment to the line where the error is reported?

Or, add `/* eslint-disable max-line-per-function */` before and  `/* eslint-enable max-line-per-function */` after the function.

**Comment by ljharb** [comment]
I can, but “i have to move code” means it’s a breaking change.

**Comment by snitin315** [comment]
Should we reopen this issue to track it for v9? The change was reverted in https://github.com/eslint/eslint/pull/15397

**Comment by mdjermanovic** [comment]
> Should we reopen this issue to track it for v9?

It seems better not to implement this change at all, for the reasons described in https://github.com/eslint/eslint/issues/15390 and in this issue.
